### PR TITLE
feat: initial implementation of metabrokerctl

### PR DESCRIPTION
**Pull request by f0rmiga**
This PR adds the initial implementation of the `metabrokerctl` CLI. The initial implementation has the following commands:
- marketplace
- provision
- deprovision
- instances
- bind
- unbind
- credentials

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,58 @@
...
+1. Start Minikube:
+
+```shell
+minikube start --cpus 4 --embed-certs
+```
+
+2. Install the CRDs:
+
+```shell
+(cd "$(git rev-parse --show-toplevel)/apis/"; make install-crd)
```
Comment (mook-as) [inline README.md]:
```suggestion
make -C "$(git rev-parse --show-toplevel)/apis/" install-crd
```
Alternatively, assume that `$PWD` is the source directory and people can figure out what you mean (and drop the `git rev-parse` magic).
Same for the other commands.

File: README.md [hunk 2]
```diff
@@ -1 +1,58 @@
...
+
+```shell
+(cd "$(git rev-parse --show-toplevel)/cli/"; make)
+```
+
+6. Provision an instance:
+
+```shell
+cd "$(git rev-parse --show-toplevel)/cli/"
+./.bin/metabrokerctl provision mypg postgres-metabroker-small
```
Comment (mook-as) [inline README.md]:
```suggestion
cli/.bin/metabrokerctl provision mypg postgres-metabroker-small
```

File: cli/Makefile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+makefile_path := $(abspath $(lastword $(MAKEFILE_LIST)))
+makefile_dir := $(patsubst %/,%,$(dir $(makefile_path)))
```
Comment (mook-as) [inline cli/Makefile:2]: Isn't that going to end up as just `$PWD` all the time?

File: cli/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	deprovisionCmd *cobra.Command,
+	instancesCmd *cobra.Command,
+	bindCmd *cobra.Command,
+	unbindCmd *cobra.Command,
+	credentialsCmd *cobra.Command,
+) *cobra.Command {
+	var kubeconfig string
+
+	cmd := &cobra.Command{
+		Use: "metabrokerctl",
```
Comment (mook-as) [inline cli/cmd/root.go:55]: I don't think this is necessary? (At least, for the root command; having less code just generally seems better, even if it's just one line.)

File: cli/cmd/bind.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+				stdout,
+				"Binding instance %q in the %q namespace to the credential %q in the %q namespace.\n",
+				instanceName, instanceNamespace, secretName, credentialNamespace,
+			)
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&credentialNamespace, "namespace", "n", "default", "The target namespace where the Credential will be created.")
```
Comment (mook-as) [inline cli/cmd/bind.go]: Hmm, I'm on the fence about making `namespace` a global flag.

File: cli/cmd/bind.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	var secretName string
+	var timeout string
+
+	cmd := &cobra.Command{
+		Use:  "bind [INSTANCE NAME] [CREDENTIAL NAME]",
+		Args: cobra.ExactArgs(2),
+		// TODO: a better description.
+		Short: "Creates a Credential object.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			instanceName := args[0]
```
Comment (mook-as) [inline cli/cmd/bind.go:56]: Do we need to do any validation? Or are we relying on the server for that?

File: cli/cmd/credentials.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+			credentials, err := clientset.
+				ServicebrokerV1alpha1().
+				Credentials(namespace).
+				List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to get credentials: %w", err)
+			}
+
+			if len(credentials.Items) > 0 {
```
Comment (mook-as) [inline cli/cmd/credentials.go:76]: Future feature request: `--output=json`.

File: cli/cmd/bind.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+			)
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&credentialNamespace, "namespace", "n", "default", "The target namespace where the Credential will be created.")
+	cmd.Flags().StringVar(&instanceNamespace, "instance-namespace", "default", "The target namespace where the Instance was created.")
+	cmd.Flags().StringVar(&secretName, "secret-name", "", "The secret name where the credentials should be created. The secret must not exist. If empty, a secret with the Credential name is created.")
+	cmd.Flags().StringVar(&timeout, "timeout", "3m0s", "The time to wait for the Credential to be created.")
```
Comment (mook-as) [inline cli/cmd/bind.go]: Why not `cmd.Flags().DurationVar()`, so we don't have to parse it ourselves?

**Review by mook-as — CHANGES_REQUESTED** [review]
Thought experiment: implement this as a series of [kubectl plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) so we can `kubectl create credential…`

… no, that probably wouldn't work well, would it, even if it does mirror the ability to do things like `kubectl create secret generic …`

File: README.md (hunk 1 above)
Comment (f0rmiga) [reply README.md]: Addressed in 48ab47a.

File: README.md (hunk 2 above)
Comment (f0rmiga) [reply README.md]: Addressed in 48ab47a.

File: cli/Makefile (hunk 3 above)
Comment (f0rmiga) [reply cli/Makefile:2]: Hmmm I don't think so, if the target is called from another Makefile.

File: cli/cmd/root.go (hunk 4 above)
Comment (f0rmiga) [reply cli/cmd/root.go:55]: This is not _necessary_ but then the help command shows empty, which is weird. Well, ok, it's wrong.

File: cli/cmd/bind.go (hunk 5 above)
Comment (f0rmiga) [reply cli/cmd/bind.go]: Oh, right... I need `LocalFlags` instead.

Comment (f0rmiga) [reply cli/cmd/bind.go]: Addressed in 55a1340.

File: cli/cmd/bind.go (hunk 6 above)
Comment (f0rmiga) [reply cli/cmd/bind.go:56]: Relying on the server.

File: cli/cmd/bind.go (hunk 5 above)
Comment (mook-as) [reply cli/cmd/bind.go]: Hahaha, I actually completely missed that and was thinking of having it on `root` instead, but I'm glad I accidentally caught a thing!

File: cli/cmd/credentials.go (hunk 7 above)
Comment (f0rmiga) [reply cli/cmd/credentials.go:76]: Added as an issue instead of TODO: https://github.com/SUSE/metabroker/issues/28.

File: cli/cmd/bind.go (hunk 8 above)
Comment (f0rmiga) [reply cli/cmd/bind.go]: Lovely! Addressed in 64d903b.

**Comment by f0rmiga** [comment]
@mook-as I don't think we get anything useful for making it a kubectl plugin.

**Review by mook-as — APPROVED** [review]
👍 
Yeah, I convinced myself that a `kubectl` plugin wasn't the way to go.

**Pull request merged**